### PR TITLE
`struct MsacAsmContextBuf`: Refactor out and `unsafe impl Send + Sync`

### PR DESCRIPTION
**Pull request by kkysen**
This refactors out `MsacAsmContext::buf_{pos,end}` to `MsacAsmContextBuf` and then `unsafe impl Send + Sync` on it since it stores pointers from the `Send + Sync` `Option<CArc<[u8]>>` in `MsacContext::asm`.  This makes `MsacContext: Send + Sync`.

### CODE REVIEW

File: src/msac.rs [hunk 1]
```diff
@@ -149,10 +149,43 @@ impl Default for Rav1dMsacDSPContext {
...
 pub type EcWin = usize;
 
+#[repr(C)]
+struct MsacAsmContextBuf {
+    pos: *const u8,
+    end: *const u8,
+}
+
+/// SAFETY: [`MsacAsmContextBuf`] is always contained in [`MsacAsmContext::buf`],
+/// which is always contained in [`MsacContext::asm`], whos [`MsacContext::data`] field
```
Comment (rinon) [inline src/msac.rs:159]:
```suggestion
/// which is always contained in [`MsacContext::asm`], whose [`MsacContext::data`] field
```

File: src/msac.rs [hunk 2]
```diff
@@ -149,10 +149,43 @@ impl Default for Rav1dMsacDSPContext {
...
+/// is what is stored in [`MsacAsmContextBuf::pos`] and [`MsacAsmContextBuf::end`].
+/// Since [`MsacContext::data`] is [`Send`], [`MsacAsmContextBuf`] is also [`Send`].
+unsafe impl Send for MsacAsmContextBuf {}
+
+/// SAFETY: [`MsacAsmContextBuf`] is always contained in [`MsacAsmContext::buf`],
+/// which is always contained in [`MsacContext::asm`], whos [`MsacContext::data`] field
```
Comment (rinon) [inline src/msac.rs:165]:
```suggestion
/// which is always contained in [`MsacContext::asm`], whose [`MsacContext::data`] field
```

File: src/msac.rs [hunk 3]
```diff
@@ -149,10 +149,43 @@ impl Default for Rav1dMsacDSPContext {
 
 pub type EcWin = usize;
 
+#[repr(C)]
+struct MsacAsmContextBuf {
```
Comment (rinon) [inline src/msac.rs:153]: Document what this must point to hold the unsafe impl invariants. Either these must both be null or they must point into the `data` field of a `MsacContext` that also contains this structure.

File: src/msac.rs [hunk 4]
```diff
@@ -149,10 +149,49 @@ impl Default for Rav1dMsacDSPContext {
 
 pub type EcWin = usize;
 
+/// # Safety
+///
+/// [`Self`] must be the first field of [`MsacAsmContext`] for asm layout purposes.
+/// And [`Self::pos`] and [`Self::end`] must be either [`ptr::null`],
+/// or [`Self::pos`] must point into (or the end of) [`MsacContext::data`],
```
Comment (rinon) [inline src/msac.rs:156]: I think it's important to point out that it must be the containing `MsacContext`, not some unrelated buffer.

**Review by rinon — APPROVED** [review]

File: src/msac.rs (hunk 4 above)
Comment (kkysen) [reply src/msac.rs:156]: Okay I'll point that out.

Comment (kkysen) [reply src/msac.rs:157]: Fixed in [`1ec6eab..6856860`](https://github.com/memorysafety/rav1d/compare/1ec6eab4e00770b6022a66294f7f3c91b3967257..6856860f24d841e011e8a69508e417699b1042c0).

**Review by rinon — APPROVED** [review]

**Pull request merged**